### ISSUE TITLE
训练代码中的问题

### ISSUE THREAD

**Issue by ge-bin-hui** [mashijie1028/GenHancer#5]
非常感谢您的工作！
`    for epoch in range(first_epoch, args.num_train_epochs):
        train_loss = 0.0
        for step, batch in enumerate(train_dataloader):
            # NOTE!!!   # accumulate super_model
            with accelerator.accumulate(super_model):
                original_img, prompts = batch['image'], batch['text']
                original_img = original_img.to(accelerator.device)
                
                with torch.no_grad():
                    x_1 = vae.encode(NORMALIZE_VAE(original_img).to(torch.float32))

                inp = prepare_clip(clip=super_model.clip_vis, original_img=NORMALIZE_CLIP(original_img).to(weight_dtype), img=x_1.to(weight_dtype))
                x_1 = rearrange(x_1, "b c (h ph) (w pw) -> b (h w) (c ph pw)", ph=2, pw=2)
                bs = original_img.shape[0]
                t = torch.sigmoid(torch.randn((bs,), device=accelerator.device) * args.scale_factor)
                x_0 = torch.randn_like(x_1).to(accelerator.device)
                x_t = (1 - t[:, None, None]) * x_1 + t[:, None, None] * x_0
                bsz = x_1.shape[0]
                guidance_vec = torch.full((x_t.shape[0],), 4, device=x_t.device, dtype=x_t.dtype)

                # Predict the noise residual and compute loss
                model_pred = super_model.dit(img=x_t.to(weight_dtype),
                                             img_ids=inp['img_ids'].to(weight_dtype),
                                             txt=inp['txt'].to(weight_dtype),
                                             txt_ids=inp['txt_ids'].to(weight_dtype),
                                             y=inp['vec'].to(weight_dtype),
                                             timesteps=t.to(weight_dtype),
                                             guidance=guidance_vec.to(weight_dtype),)

                loss = F.mse_loss(model_pred.float(), (x_0 - x_1).float(), reduction="mean")

                # Gather the losses across all processes for logging (if we use distributed training).
                avg_loss = accelerator.gather(loss.repeat(args.train_batch_size)).mean()
                print(f"avg_loss{accelerator.process_index}:{avg_loss}")
                train_loss += avg_loss.item() / args.gradient_accumulation_steps
                print(f"train_loss{accelerator.process_index}:{train_loss}")

                # Backpropagate
                accelerator.backward(loss)
                if accelerator.sync_gradients:
                    accelerator.clip_grad_norm_(super_model.parameters(), args.max_grad_norm)
                optimizer.step()
                lr_scheduler.step()
                optimizer.zero_grad()

            # Checks if the accelerator has performed an optimization step behind the scenes
            if accelerator.sync_gradients:
                progress_bar.update(1)
                global_step += 1
                accelerator.log({"train_loss": train_loss}, step=global_step)
                train_loss = 0.0

                if global_step % args.checkpointing_steps == 0:
                    if accelerator.is_main_process:
                        unwrapped_super_model = accelerator.unwrap_model(super_model)
                        # save dit ckpts
                        save_path_dit = os.path.join(args.output_dir, f"checkpoint-dit-{global_step}.bin")
                        torch.save(deepcopy(unwrapped_super_model.dit).state_dict(), save_path_dit)
                        # save project clip ckpts
                        save_path_project_clip = os.path.join(args.output_dir, f"checkpoint-project-clip-{global_step}.bin")
                        torch.save(deepcopy(unwrapped_super_model.clip_vis.project_clip).state_dict(), save_path_project_clip)
                        # save project t5 ckpts
                        save_path_project_t5 = os.path.join(args.output_dir, f"checkpoint-project-t5-{global_step}.bin")
                        torch.save(deepcopy(unwrapped_super_model.clip_vis.project_t5).state_dict(), save_path_project_t5)
                        # save optimizer state
                        save_path_optimizer = os.path.join(args.output_dir, f"optimizer-state-{global_step}.bin")
                        torch.save(optimizer.state_dict(), save_path_optimizer)

                        logger.info(f"Saved ckpts to {save_path_dit}, {save_path_project_clip} and {save_path_project_t5}")

            logs = {"step_loss": loss.detach().item(), "lr": lr_scheduler.get_last_lr()[0]}
            progress_bar.set_postfix(**logs)

            if global_step >= args.max_train_steps:
                if accelerator.is_main_process:
                    unwrapped_super_model = accelerator.unwrap_model(super_model)
                    # save dit ckpts
                    save_path_dit = os.path.join(args.output_dir, f"checkpoint-dit-{global_step}.bin")
                    torch.save(deepcopy(unwrapped_super_model.dit).state_dict(), save_path_dit)
                    # save project clip ckpts
                    save_path_project_clip = os.path.join(args.output_dir, f"checkpoint-project-clip-{global_step}.bin")
                    torch.save(deepcopy(unwrapped_super_model.clip_vis.project_clip).state_dict(), save_path_project_clip)
                    # save project t5 ckpts
                    save_path_project_t5 = os.path.join(args.output_dir, f"checkpoint-project-t5-{global_step}.bin")
                    torch.save(deepcopy(unwrapped_super_model.clip_vis.project_t5).state_dict(), save_path_project_t5) 

                break`
在上述您提供的代码中，我发现了一些问题。在多卡并行时，每张gpu所使用的数据是相同的，即没有进行多卡训练。此外，在梯度更新部分也存在一些问题，即没有将optimizer.step()，lr_scheduler.step()，optimizer.zero_grad()写入if accelerator.sync_gradients: 判断语句中，这会导致梯度不会累积。
